Source/Plugins/Processing/ProcessingFilters/ConvertColorToGrayScale.cpp

```diff
@@ -72,106 +79,104 @@ class LuminosityImpl
 #endif
-private:
-  uint8_t* m_ImageData;
-  uint8_t* m_FlatImageData;
-  FloatVec3_t m_ColorWeights;
-  size_t m_NumComp;
+
 };
 
 /**
- * @brief The ConvertColorToGrayScaleImpl class implements a threaded algorithm
+ * @brief The ParallelWrapper class implements a threaded algorithm
  * that flattens an RGB array into a grayscale array
  */
-class AverageImpl
+class LightnessImpl
 {
+private:
+  uint8_t* m_ImageData;
+  uint8_t* m_FlatImageData;
+  size_t m_NumComp;
+
 public:
-  AverageImpl(uint8_t* data, uint8_t* newdata, size_t comp)
+  LightnessImpl(uint8_t* data, uint8_t* newdata, size_t comp)
   : m_ImageData(data)
   , m_FlatImageData(newdata)
   , m_NumComp(comp)
   {
   }
-  virtual ~AverageImpl() = default;
+  ~LightnessImpl()  = default;
 
   void convert(size_t start, size_t end) const
   {
     for(size_t i = start; i < end; i++)
     {
-      uint16_t sum = 0;
+      uint8_t maxRGB = 0;
+      uint8_t minRGB = 255;
+
       for(size_t j = 0; j < 3; j++)
       {
-        sum += static_cast<uint16_t>(m_ImageData[m_NumComp * i + j]);
+        if(m_ImageData[m_NumComp * i + j] > maxRGB)
+        {
+          maxRGB = m_ImageData[m_NumComp * i + j];
+        }
+        if(m_ImageData[m_NumComp * i + j] < minRGB)
+        {
+          minRGB = m_ImageData[m_NumComp * i + j];
+        }
       }
-
-      sum = sum / 3;
-      m_FlatImageData[i] = static_cast<uint8_t>(sum);
+      m_FlatImageData[i] = static_cast<uint8_t>(roundf((maxRGB + minRGB) / 2.0f));
     }
   }
 
+
 #ifdef SIMPL_USE_PARALLEL_ALGORITHMS
-  void operator()(const tbb::blocked_range<size_t>& r) const
-  {
-    convert(r.begin(), r.end());
-  }
+  void operator()(const tbb::blocked_range<size_t>& r) const { convert(r.begin(), r.end()); }
 #endif
-private:
-  uint8_t* m_ImageData;
-  uint8_t* m_FlatImageData;
-  size_t m_NumComp;
 };
 
 /**
  * @brief The ConvertColorToGrayScaleImpl class implements a threaded algorithm
  * that flattens an RGB array into a grayscale array
  */
-class LightnessImpl
+class AverageImpl
 {
 public:
-  LightnessImpl(uint8_t* data, uint8_t* newdata, size_t comp)
+  AverageImpl(uint8_t* data, uint8_t* newdata, size_t comp)
   : m_ImageData(data)
   , m_FlatImageData(newdata)
   , m_NumComp(comp)
   {
   }
-  virtual ~LightnessImpl() = default;
+  virtual ~AverageImpl() = default;
 
   void convert(size_t start, size_t end) const
   {
     for(size_t i = start; i < end; i++)
     {
-      uint8_t maxRGB = 0;
-      uint8_t minRGB = 255;
-
+      uint16_t sum = 0;
       for(size_t j = 0; j < 3; j++)
       {
-        if(m_ImageData[m_NumComp * i + j] > maxRGB)
-        {
-          maxRGB = m_ImageData[m_NumComp * i + j];
-        }
-        if(m_ImageData[m_NumComp * i + j] < minRGB)
-        {
-          minRGB = m_ImageData[m_NumComp * i + j];
-        }
+        sum += static_cast<uint16_t>(m_ImageData[m_NumComp * i + j]);
       }
-      m_FlatImageData[i] = static_cast<uint8_t>((maxRGB + minRGB) / 2.0);
+
+      sum = sum / 3;
```

Comment:
`#include <algorithm>`
`const size_t compSize = 3;`
`auto startIter= std::begin(m_ImageData) + (start * compSize);`
`auto endIter = std::end(m_ImageData) + (end * compSize);`
`uint16_t sum = std::accumulate(startIter, endIter, 0) / compSize;`

Reply:
Forgot `#include <iterator>` and perhaps `std::cbegin` and `std::cend` would be better since m_ImageData isn't being changed.